### PR TITLE
Add retreive vote functionality

### PR DESCRIPTION
**Pull request by tiborv**
#92

### CODE REVIEW

**Comment by ekmartin** [comment]
Kanskje kalle det kvittering i stedet for hash i frontenden? :)

File: public/js/controllers/electionCtrl.js [hunk 1]
```diff
@@ -36,6 +36,9 @@ function($scope, electionService, alertService, voteService, socketIOService) {
             .success(function(vote) {
                 $scope.activeElection = null;
                 alertService.addSuccess('Takk for din stemme!');
+                localStorageService.set('voteHash', vote.hash);
+                alertService.addSuccess(localStorageService.get('voteHash'));
```
Comment (ekmartin) [inline public/js/controllers/electionCtrl.js]: Delete this?

File: public/js/controllers/retrieveVoteCtrl.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+angular.module('voteApp').controller('retrieveVoteController',
+['$scope', 'alertService', 'voteService', 'localStorageService',
+function($scope, alertService, voteService, localStorageService) {
+    $scope.vote = { hash: localStorageService.get('voteHash')};
+
+    $scope.retreiveVote =  function(vote) {
+        voteService.retrieve(vote.hash)
+            .success(function(vote) {
+                var msg =
+                    'Funnet! Stemme med hash "' + vote.hash +
```
Comment (ekmartin) [inline public/js/controllers/retrieveVoteCtrl.js]: This should probably be done in the template, and not by adding an alert. Can use `ng-if` to choose if it should be shown or not, and bind the vote to scope (`$scope.vote = vote`).

File: test/api/vote.test.js [hunk 3]
```diff
@@ -243,7 +243,7 @@ describe('Vote API', function() {
                         if (err) return done(err);
 
                         var receivedVote = res.body;
-                        receivedVote.alternative.should.equal(String(vote.alternative));
+                        receivedVote.alternative._id.should.equal(String(vote.alternative));
```
Comment (ekmartin) [inline test/api/vote.test.js]: Should probably test the populated fields too? I.e. `vote.alternative.election` etc.

File: features/election.feature [hunk 4]
```diff
@@ -21,6 +21,22 @@ Feature: Election
     When I vote on an election
     Then I see alert "Takk for din stemme!"
 
+  Scenario: Retrieve vote with valid hash
+    Given There is an active election
+    And I have voted on the election
+    And I am on the retrieve vote page
+    When I submit the form
+    Then I see "Kvittering:"
```
Comment (ekmartin) [inline features/election.feature]: Don't think there's any point in checking the titles, should just verify that the actual values are correct. Could use `element(by.binding('vote.fieldGoesHere')` to match `{{vote.fieldGoesHere}}`.

File: features/step_definitions/electionSteps.js [hunk 5]
```diff
@@ -18,6 +18,22 @@ module.exports = function() {
...
+        callback();
+    });
+
+    this.Given(/^I enter a random string as vote hash$/, function(callback) {
+        element(by.model('voteHash')).sendKeys('LAALLALALA');
+        callback();
+    });
+
+    this.When(/^I submit the form$/, function(callback) {
+        var button = element(by.css('button'));
```
Comment (ekmartin) [inline features/step_definitions/electionSteps.js]: This'll fail if there's a button outside the form, should probably use `element(by.css('form'))` or `element(by.tagName('form'))` instead.

File: public/js/controllers/retrieveVoteCtrl.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+                $scope.vote = vote;
+            })
+            .error(function(error) {
+                switch (error.name) {
+                    case 'NotFoundError':
+                        alertService.addError('En stemme med denne kvitteringen ble ikke funnet.');
+                        $scope.vote = null;
+                        break;
+                    default:
+                        alertService.addError('Noe gikk galt, vennligst prøv igjen.');
```
Comment (ekmartin) [inline public/js/controllers/retrieveVoteCtrl.js]: Can just use `alertService.addError()` and it'll use the default message ("Noe gikk galt!").

File: features/step_definitions/electionSteps.js [hunk 7]
```diff
@@ -18,6 +18,22 @@ module.exports = function() {
         this.election.save(callback);
     });
 
+    this.Given(/^I am on the retrieve vote page$/, function(callback) {
+        browser.get('/retrieve');
+        callback();
+    });
+
+    this.Given(/^I enter a random string as vote hash$/, function(callback) {
```
Comment (ekmartin) [inline features/step_definitions/electionSteps.js]: If #103 gets merged first you can use https://github.com/webkom/vote/pull/103/files#diff-f42390971b8bdd5b2f707a60adbf8b6fR89 - `I fill in model "voteHash" with "notahash"`

**Comment by ekmartin** [comment]
Maybe we should have one test to make sure that localStorage works and one that fills in the hash (correctly) manually?

**Comment by ekmartin** [comment]
Updated the styling and frontend tests now, like we talked about :)
It also deletes the voteHash from localStorage on logout now.

**Comment by JuulArthur** [comment]
Good job!

**Pull request merged**